### PR TITLE
Removed connector dependencies from Store and ReSwift, added tests 

### PR DESCRIPTION
**Pull request by santaarturio**
## Description

Swapped connector dependencies from Store and ReSwift with dependencies from Provider, added tests & Mocks for Connectors

### Types of changes

What types of changes does your code introduce to the Connect app?

- [x] Tech debt (non-breaking change which adds functionality)

### CODE REVIEW

File: TopMovies/Connector/CategoryVCConnector.swift [hunk 1]
```diff
@@ -5,19 +5,20 @@
 //  Created by Macbook Pro  on 17.02.2021.
 //
 
-import Foundation
-
-final class CategoryVCConnector: BaseConnector<MoviesCategoryVCProps> {
+final class CategoryVCConnector<Provider: StoreProviderProtocol>: BaseConnector<MoviesCategoryVCProps, Provider>
+where Provider.ExpectedStateType == MainState {
   var categoryId: MovieCategory.ID?
```
Comment (balavor) [inline TopMovies/Connector/CategoryVCConnector.swift]: `var` -> `let`
`MovieCategory.ID?` -> `MovieCategory.ID`

File: TopMovies/Connector/TopMoviesConnector.swift [hunk 2]
```diff
@@ -5,17 +5,19 @@
...
-  required init(updateProps: @escaping (TopMoviesProps) -> Void) {
-    super.init(updateProps: updateProps)
+  typealias StateUpdate = (Provider.ExpectedStateType) -> Void
+  override init(updateProps: @escaping (TopMoviesProps) -> Void,
+                provider: (@escaping StateUpdate) -> Provider) {
     
-    mainStore.dispatch(RequestMovieCategoriesAction())
+    super.init(updateProps: updateProps,
+               provider: provider)
+    self.provider.dispatch(RequestMovieCategoriesAction())
```
Comment (balavor) [inline TopMovies/Connector/TopMoviesConnector.swift:20]: You can omit `self`.

File: TopMovies/Resources/Router.swift [hunk 3]
```diff
@@ -43,9 +45,12 @@ class Router {
...
     let categoryVc = MoviesCategoryVC()
     categoryVc.configureConnectionWith(
-      connector: CategoryVCConnector(categoryId: categoryId) { [unowned categoryVc] (props) in
-        categoryVc.connect(props: props)
-      })
+      connector: CategoryVCConnector(
+        categoryId: categoryId,
+        updateProps: { [unowned categoryVc] (props) in
+          categoryVc.connect(props: props)
+        },
```
Comment (balavor) [inline TopMovies/Resources/Router.swift]: Is it possible to move that inside of the init of the `BaseConnector`?
Looks like a duplication. Moreover, that code is required to make connector work with vc.

File: TopMovies/View/CategoryVC/MoviesCategoryVC.swift [hunk 4]
```diff
@@ -20,7 +20,10 @@ struct MoviesCategoryVCProps {
 
 // MARK: - VC class
 final class MoviesCategoryVC: UIViewController, PropsConnectable {
```
Comment (balavor) [inline TopMovies/View/CategoryVC/MoviesCategoryVC.swift]: Just mentioned that... Let's stick with a single name across the project.
It's either `***VC: UIViewController`, or `***ViewController: UIViewController`

**Review by balavor — CHANGES_REQUESTED** [review]

File: TopMovies/View/CategoryVC/MoviesCategoryVC.swift (hunk 4 above)
Comment (santaarturio) [reply TopMovies/View/CategoryVC/MoviesCategoryVC.swift]: got it, thanks!

File: TopMovies/Connector/TopMoviesConnector.swift (hunk 2 above)
Comment (santaarturio) [reply TopMovies/Connector/TopMoviesConnector.swift:20]: > You can omit `self`.

I can't, cause self.provider and provider in init(provider:) have different types but the same name

File: TopMovies/Resources/Router.swift (hunk 3 above)
Comment (santaarturio) [reply TopMovies/Resources/Router.swift]: > Is it possible to move that inside of the init of the `BaseConnector`?
> Looks like a duplication. Moreover, that code is required to make connector work with vc.

How about Delegate pattern ?)

File: TopMovies/Connector/PropsConnectableProtocol.swift [hunk 5]
```diff
@@ -1,15 +0,0 @@
...
-//
-
-import Foundation
-
-protocol PropsConnectable {
-  associatedtype Props
-  var propsConnector: BaseConnector<Props>? { get set }
-  func configureConnectionWith(connector: BaseConnector<Props>)
-  func connect(props: Props)
-}
```
Comment (balavor) [inline TopMovies/Connector/PropsConnectableProtocol.swift:15]: Even though we have a BaseVC, or BaseConnector, let them still follow the protocol, cz the base class is not a pure interface.
It provides his inheritors some functionality.

File: TopMovies/Basic/BaseVC.swift [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+import UIKit
+
+class BaseVC<Props, Provider: StoreProviderProtocol>: UIViewController {
+  typealias Connector = BaseConnector<Props, Provider>
+  private(set) var propsConnector: Connector?
+  
+  typealias PropsUpdate = (Props) -> Void
+  final func configureConnectionWith(connector: (@escaping PropsUpdate) -> Connector) {
+    propsConnector = connector { [unowned self] props in connect(props: props) }
+  }
```
Comment (balavor) [inline TopMovies/Basic/BaseVC.swift]: We could avoid `final` to allow override that method.
For example, if you will have some child view controllers that will require connectors from main connector:
```swift
class SomeConnector: BaseConnector {
  let connectors: [BaseConnector]
}

class SomeVC: BaseVC {
  let controllers: [BaseVC] 
  let connector: BaseConnector
  
  override func configureConnection(with connector: (@escaping PropsUpdate) -> Connector) {
    super. configureConnection(with: connector)
    zip(self.controllers, connector.connectors) { vc, connector in
      // vc configured with connector only when routing starts
    }
  }
}
```

It's just an example of what we could do, no need to adopt smth like that right now :)

File: TopMovies/Basic/BaseVC.swift [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+//
+
+import UIKit
+
+class BaseVC<Props, Provider: StoreProviderProtocol>: UIViewController {
+  typealias Connector = BaseConnector<Props, Provider>
+  private(set) var propsConnector: Connector?
+  
+  typealias PropsUpdate = (Props) -> Void
+  final func configureConnectionWith(connector: (@escaping PropsUpdate) -> Connector) {
```
Comment (balavor) [inline TopMovies/Basic/BaseVC.swift]:
```swift
 func configureConnection(with connector: (@escaping PropsUpdate) -> Connector) 
```

**Review by balavor — CHANGES_REQUESTED** [review]

**Review by balavor — APPROVED** [review]

**Pull request merged**